frame/treasury/src/lib.rs

```diff
@@ -278,71 +222,162 @@ decl_module! {
-		/// # </weight>
-		#[weight = (T::WeightInfo::approve_proposal(), DispatchClass::Operational)]
-		pub fn approve_proposal(origin, #[compact] proposal_id: ProposalIndex) {
+		#[pallet::weight(T::WeightInfo::approve_proposal())]
+		pub fn approve_proposal(
+			origin: OriginFor<T>,
+			#[pallet::compact] proposal_id: ProposalIndex,
+		) -> DispatchResultWithPostInfo {
 			T::ApproveOrigin::ensure_origin(origin)?;
 
 			ensure!(<Proposals<T, I>>::contains_key(proposal_id), Error::<T, I>::InvalidIndex);
-			Approvals::<I>::append(proposal_id);
+			<Approvals<T, I>>::append(proposal_id);
+			Ok(().into())
 		}
+	}
 
-		/// # <weight>
-		/// - Complexity: `O(A)` where `A` is the number of approvals
-		/// - Db reads and writes: `Approvals`, `pot account data`
-		/// - Db reads and writes per approval:
-		///   `Proposals`, `proposer account data`, `beneficiary account data`
-		/// - The weight is overestimated if some approvals got missed.
-		/// # </weight>
-		fn on_initialize(n: T::BlockNumber) -> Weight {
-			// Check to see if we should spend some funds!
-			if (n % T::SpendPeriod::get()).is_zero() {
-				Self::spend_funds()
-			} else {
-				0
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	#[pallet::metadata(T::AccountId = "AccountId", BalanceOf<T, I> = "Balance")]
+	pub enum Event<T: Config<I>, I: 'static = ()> {
+		/// New proposal. \[proposal_index\]
+		Proposed(ProposalIndex),
+		/// We have ended a spend period and will now allocate funds. \[budget_remaining\]
+		Spending(BalanceOf<T, I>),
+		/// Some funds have been allocated. \[proposal_index, award, beneficiary\]
+		Awarded(ProposalIndex, BalanceOf<T, I>, T::AccountId),
+		/// A proposal was rejected; funds were slashed. \[proposal_index, slashed\]
+		Rejected(ProposalIndex, BalanceOf<T, I>),
+		/// Some of our funds have been burnt. \[burn\]
+		Burnt(BalanceOf<T, I>),
+		/// Spending has finished; this is the amount that rolls over until next spend.
+		/// \[budget_remaining\]
+		Rollover(BalanceOf<T, I>),
+		/// Some funds have been deposited. \[deposit\]
+		Deposit(BalanceOf<T, I>),
+	}
+
+	/// Old name generated by `decl_event`.
+	// TODO :: Have to re-check is "Note message" required ?
+	// #[deprecated(note = "use `Event` instead")]
+	pub type RawEvent<T, I = ()> = Event<T, I>;
+
+	#[pallet::error]
+	pub enum Error<T, I = ()> {
+		/// Proposer's balance is too low.
+		InsufficientProposersBalance,
+		/// No proposal or bounty at that index.
+		InvalidIndex,
+	}
+
+	/// Number of proposals that have been made.
+	#[pallet::storage]
+	#[pallet::getter(fn proposal_count)]
+	pub type ProposalCount<T: Config<I>, I: 'static = ()> = StorageValue<_, ProposalIndex, ValueQuery>;
+
+	/// Proposals that have been made.
+	#[pallet::storage]
+	#[pallet::getter(fn proposals)]
+	pub type Proposals<T: Config<I>, I: 'static = ()> = StorageMap<
+		_,
+		Twox64Concat,
+		ProposalIndex,
+		Option<Proposal<T::AccountId, BalanceOf<T, I>>>,
+		ValueQuery,
```

Comment:
comparing the metadata should show these kind of error.
```suggestion
		Proposal<T::AccountId, BalanceOf<T, I>>,
		OptionQuery,
```